designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,784 @@
+    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
+1. When a glob pattern is passed directly (such as `eslint src/*.js`):
+    1. ESLint expands the glob pattern to get a list of files.
+    1. Each file is checked individually as in step 1.
+1. When a directory is passed directly (such as `eslint src`):
+    1. The directory is converted into a glob pattern by appending `**/*` to the directory (such as `src` becomes `src/**/*`).
+    1. The glob pattern is checked as in step 2.
+1. When a relative directory is passed directly (such as `eslint .`):
+    1. The relative directory pattern is resolved to a full directory name.
+    1. The glob pattern is checked as in step 3.
+
+**Note:** ESLint will continue to ignore `node_modules` by default.
+
+### Rename `--no-eslintrc` to `--no-config-file` and `useEslintrc` to `useConfigFile`
+
+Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.
+
+## Documentation
+
+This will require extensive documentation changes and an introductory blog post.
+
+At a minimum, these pages will have to be updated (and rewritten):
+
+* https://eslint.org/docs/user-guide/getting-started#configuration
+* https://eslint.org/docs/user-guide/configuring
+* https://eslint.org/docs/developer-guide/working-with-plugins#configs-in-plugins
+* https://eslint.org/docs/developer-guide/shareable-configs
+
+## Drawbacks
+
+As with any significant change, there are some significant drawbacks:
+
+1. We'd need to do a phased rollout (see Backwards Compatibility Analysis) to minimize impact on users. That means maintaining two configuration systems for a while, increasing maintenance overhead and complexity of the application.
+1. Getting everyone to convert to `eslint.config.js` format would be a significant stress on the community that could cause some resentment.
+1. We can no longer enforce naming conventions for plugins and shareable configs. This may make it more difficult for users to find compatible npm packages.
+1. Creating configuration files will be more complicated.
+1. The `--print-config` option becomes less useful because we can't output objects in a meaningful way.
+1. People depending on environments may find this change particularly painful.
+
+## Backwards Compatibility Analysis
+
+The intent of this proposal is to replace the current `.eslintrc` format, but can be implemented incrementally so as to cause as little disturbance to ESLint users as possible
+
+In the first phase, I envision this:
```

Comment:
I don't think labeling it as experimental is helpful and might work counter to what we need by dissuading people from trying it out. If we agree that this is a good design to implement, then we should commit to it. The idea of phasing this format in and phasing out the old format is to catch any usability or other issues and address them, just like any other feature. It's unlikely people will jump immediately to the new format unless it's solving an immediate need, so we should have plenty of time before we remove the old functionality.

Reply:
Agreed that the goal is to catch usability issues, and that we should commit to this design before implementing it. I'm saying that if fixing a usability issue requires a breaking change to the new format, it could complicate adoption given our typical scheduling for major releases. By marking the new config format as experimental, we would be able to fix usability issues without waiting a year for the next major release, and then requiring users who migrated early to migrate again.

Reply:
Again, I just don't see this as being any different from any other feature we implement.

Reply:
@not-an-aardvark, I don't think we need any "experimental" thing. It's different enough. It's opt-in anyway. If you choose to use the `eslint.config.js` then you know what you're doing and that it's experimental. We don't need to have "--experimental" for or such things. Or what you mean by "experimental" more specifically?